libproxy/proxy.cpp

```diff
@@ -56,398 +37,57 @@ class proxy_factory {
-	this->mutex = CreateMutex(NULL, false, NULL);
-	WSADATA wsadata;
-	WORD vers = MAKEWORD(2, 2);
-	WSAStartup(vers, &wsadata);
-#else
-	pthread_mutex_init(&this->mutex, NULL);
-#endif
-
-	lock();
-
-	this->pac    = NULL;
-	this->pacurl = NULL;
-	this->wpad   = false;
 
-	// Register our types
-	this->mm.register_type<config_extension>();
-	this->mm.register_type<ignore_extension>();
-	this->mm.register_type<network_extension>();
-	this->mm.register_type<pacrunner_extension>();
-	this->mm.register_type<wpad_extension>();
-
-	// Load builtin modules
-	for (int i=0 ; builtin_modules[i] ; i++)
-		this->mm.load_builtin(builtin_modules[i]);
-
-	// Load all modules
-	module_dir = getenv("PX_MODULE_PATH");
-	if (!module_dir)
-		module_dir = MODULEDIR;
-	this->mm.load_dir(module_dir);
-	this->mm.load_dir(module_dir, false);
-
-	this->debug  = (getenv("_PX_DEBUG") != NULL);
-
-	unlock();
+  this->proxy = g_dbus_proxy_new_for_bus_sync (G_BUS_TYPE_SESSION,
+                                         G_DBUS_PROXY_FLAGS_NONE,
+                                         NULL, /* GDBusInterfaceInfo */
+                                         "org.libproxy.proxy",
+                                         "/org/libproxy/proxy",
+                                         "org.libproxy.proxy",
+                                         NULL, /* GCancellable */
+                                         this->error);
 }
 
 proxy_factory::~proxy_factory() {
-	lock();
-
-	if (this->pac) delete[] this->pac;
-	if (this->pacurl) delete this->pacurl;
-	
-	unlock();
 
-#ifdef WIN32
-	WSACleanup();
-	ReleaseMutex(this->mutex);
-#else
-	pthread_mutex_destroy(&this->mutex);
-#endif
 }
 
 
 vector<string> proxy_factory::get_proxies(const string &realurl) {
 	vector<string>             response;
 
-	// Check to make sure our url is valid
-	if (!url::is_valid(realurl))
-		goto do_return;
 
-	lock();
+GVariant* result;
 
-	// Let trap and forward exceptions so we don't deadlock
-	try {
-		vector<url> configs;
-		string ignore;
-		url dst(realurl);
+result = g_dbus_proxy_call_sync (proxy,
+                        "query",
+                        g_variant_new("(s)", realurl.c_str()),
+                        G_DBUS_CALL_FLAGS_NONE,
+                        -1,
+                        NULL,
+                        this->error);
 
-		check_network_topology();
-		get_config(dst, configs, ignore);
-		
-		if (debug) cerr << "Config is: " << endl;
 
-		for (vector<url>::iterator i=configs.begin() ; i != configs.end() ; i++) {
-			url confurl(*i);
+  GVariantIter *iter;
+  gchar *str;
 
-			if (debug) cerr << "\t" << confurl.to_string() << endl;
-
-			if (expand_wpad(confurl) || expand_pac(confurl)) {
-				run_pac(dst, confurl, response);
-			} else {
-				clear_cache();
-				response.push_back(confurl.to_string());
-			}
-		}
-
-		unlock();
-	} catch (exception &e) {
-		unlock();
-		throw e;
-	}
+  g_variant_get (result, "(as)", &iter);
+  while (g_variant_iter_loop (iter, "s", &str))
```

Comment:
I think str is leaked here? It probably needs to be freed with g_free() after it gets copied into the response vector?